src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -0,0 +1,1664 @@
+    m_image_cache.aio_discard(offset, length, skip_partial_discard, on_finish);
+  } else {
+    m_image_writeback.aio_discard(offset, length, skip_partial_discard, on_finish);
+  }
+  /*
+  if (!is_block_aligned({{offset, length}})) {
+    // For clients that don't use LBA extents, re-align the discard request
+    // to work with the cache
+    ldout(cct, 20) << "aligning discard to block size" << dendl;
+
+    // TODO: for non-aligned extents, invalidate the associated block-aligned
+    // regions in the cache (if any), send the aligned extents to the cache
+    // and the un-aligned extents directly to back to librbd
+  }
+
+  // TODO invalidate discard blocks until writethrough/back support added
+  C_Gather *ctx = new C_Gather(cct, on_finish);
+  Context *invalidate_done_ctx = ctx->new_sub();
+
+  m_image_writeback.aio_discard(offset, length, skip_partial_discard, ctx->new_sub());
+
+  ctx->activate();
+
+  Context *invalidate_ctx = new FunctionContext(
+    [this, offset, length, invalidate_done_ctx](int r) {
+      invalidate({{offset, length}}, invalidate_done_ctx);
+    });
+  flush(invalidate_ctx);
+  */
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::aio_flush(Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "on_finish=" << on_finish << dendl;
+
+  {
+    RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
+    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+      on_finish->complete(-EROFS);
+      return;
+    }
+  }
+
+  if (!m_flush_seen) {
+    Mutex::Locker locker(m_lock);
+    if (!m_flush_seen) {
+      ldout(cct, 5) << "flush seen" << dendl;
+      m_flush_seen = true;
+      if (!m_persist_on_flush && m_persist_on_write_until_flush) {
+	m_persist_on_flush = true;
+	ldout(cct, 5) << "now persisting on flush" << dendl;
+      }
+    }
+  }
+  
+  if (m_image_ctx.persistent_cache_enabled) {
+    m_image_cache.aio_flush(on_finish);
+  } else {
+    m_image_writeback.aio_flush(on_finish);
+  }
+  /*
+  Context *ctx = new FunctionContext(
+    [this, on_finish](int r) {
+      if (r < 0) {
+        on_finish->complete(r);
+      }
+      m_image_writeback.aio_flush(on_finish);
+    });
+
+  flush(ctx);
+  */
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::aio_writesame(uint64_t offset, uint64_t length,
+                                      bufferlist&& bl, int fadvise_flags,
+                                      Context *on_finish) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "offset=" << offset << ", "
+                 << "length=" << length << ", "
+                 << "data_len=" << bl.length() << ", "
+                 << "on_finish=" << on_finish << dendl;
+  {
+    RWLock::RLocker snap_locker(m_image_ctx.snap_lock);
+    if (m_image_ctx.snap_id != CEPH_NOSNAP || m_image_ctx.read_only) {
+      on_finish->complete(-EROFS);
+      return;
+    }
+  }
+
+  if (m_image_ctx.persistent_cache_enabled) {
```

Comment:
Similar use-case as compare-and-write and should probably just be a passthrough or initially flagged as non-supported by returning an error.